kv/txn.go

```diff
@@ -19,21 +19,269 @@ import (
 	"errors"
 	"math"
 	"math/rand"
+	"sync"
+	"sync/atomic"
 	"time"
 
 	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/parser/terror"
 	"github.com/pingcap/tidb/util/logutil"
+	"github.com/tikv/client-go/v2/oracle"
 	"go.uber.org/zap"
 )
 
+const (
+	chanBufferSize             = 256
+	undeletedStartTsBufferSize = 16
+	// TimeToPrintLongTimeInternalTxn is the duration if the internal transaction lasts more than it, TiDB prints
+	// a log message
+	TimeToPrintLongTimeInternalTxn = time.Minute
+)
+
+var globalInnerTxnTsBox atomic.Value
+
+type innerTxnStartTsBox struct {
+	wg                  sync.WaitGroup
+	chanToStoreStartTS  chan uint64
+	chanToDeleteStartTS chan uint64
+
+	innerTxnTsBoxRun          atomic.Value
+	innerTxnTsBoxRoutineCount int32
+
+	innerTSLock        sync.Mutex
+	innerTxnStartTsMap map[uint64]uint64
+
+	// `storeInnerTxnStartTsLoop` and `deleteInnerTxnStartTsLoop` run asynchronously,
+	// It can't ensure `storeInnerTxnStartTsLoop` receives startTS before `deleteInnerTxnStartTsLoop`,
+	// though `RunInNewTxn`send startTS to `storeInnerTxnStartTsLoop` firstly. If `deleteInnerTxnStartTsLoop`
+	// recevied startTS before `storeInnerTxnStartTsLoop`, the `startTS` couldn't be deleted from `innerTxnStartTsMap`,
+	// and GC safepoint can't be advanced properly.
+	undeletedTsLock        sync.Mutex
```

Comment:
Why not use a single channel and goroutine to process 2 types of task - put and delete?

Reply:
I worry about that it maybe cause a message block. I removed these code and don't use channel now